client/src/storage/device.rs

```diff
@@ -247,14 +673,336 @@ impl SenderKeyStore for DeviceSenderKeyStore {
+            "#,
+            addr,
+            rec,
+            rec
+        )
+        .execute(&self.pool)
+        .await
+        .map(|_| ())
+        .map_err(|err| SignalProtocolError::InvalidArgument(format!("{}", err)))
     }
 
     async fn load_sender_key(
         &mut self,
         sender: &ProtocolAddress,
         distribution_id: Uuid,
     ) -> Result<Option<SenderKeyRecord>, SignalProtocolError> {
-        todo!()
+        let addr = format!("{}:{}", sender, distribution_id);
+
+        match sqlx::query!(
+            r#"
+            SELECT
+                sender_key_record
+            FROM
+                DeviceSenderKeyStore
+            WHERE
+                address = ?
+            "#,
+            addr
+        )
+        .fetch_one(&self.pool)
+        .await
+        {
+            Ok(row) => SenderKeyRecord::deserialize(
+                BASE64_STANDARD
+                    .decode(row.sender_key_record)
+                    .map_err(|err| SignalProtocolError::InvalidArgument(format!("{}", err)))?
+                    .as_slice(),
+            )
+            .map(|res| Some(res)),
+            Err(_) => Ok(None),
+        }
+    }
+}
+
+#[cfg(test)]
+mod device_protocol_test {
+    use common::utils::time_now;
+    use libsignal_core::{Aci, ProtocolAddress};
+    use libsignal_protocol::{
+        kem::KeyType, Direction, GenericSignedPreKey, IdentityKeyPair, IdentityKeyStore, KeyPair,
+        KyberPreKeyId, KyberPreKeyRecord, KyberPreKeyStore, PreKeyId, PreKeyRecord, PreKeyStore,
+        SessionRecord, SessionStore, SignedPreKeyId, SignedPreKeyRecord, SignedPreKeyStore,
+    };
+    use rand::rngs::OsRng;
+    use sqlx::{sqlite::SqlitePoolOptions, SqlitePool};
+    use uuid::Uuid;
+
+    use crate::storage::device::{
+        DeviceIdentityKeyStore, DeviceKyberPreKeyStore, DevicePreKeyStore, DeviceSessionStore,
+        DeviceSignedPreKeyStore,
+    };
+
+    async fn connect() -> SqlitePool {
+        dotenv::dotenv().ok();
+        let db_url =
+            std::env::var("DATABASE_URL").expect("Expected to read database url from .env file");
+        let pool = SqlitePoolOptions::new()
+            .connect(&db_url)
+            .await
+            .expect("Could not connect to database");
+        sqlx::migrate!("client_db/migrations")
+            .run(&pool)
+            .await
+            .unwrap();
+
+        pool
+    }
+
+    #[tokio::test]
+    async fn save_and_get_identity_test() {
+        let key_pair = IdentityKeyPair::generate(&mut OsRng);
+        let mut device_identity_key_store =
+            DeviceIdentityKeyStore::create(key_pair, 1, connect().await)
+                .await
+                .unwrap();
+        let address = ProtocolAddress::new(Aci::from(Uuid::new_v4()).service_id_string(), 1.into());
+        let other_key_pair = IdentityKeyPair::generate(&mut OsRng);
+        let new_other_key_pair = IdentityKeyPair::generate(&mut OsRng);
```

Comment:
Cant we add functions in test_utils that helps with these kinds of things? We have an equvalent function for protocol address in server test_utils/user.rs